awscli/customizations/wizard/core.py

```diff
@@ -24,38 +24,42 @@ def __init__(self, step_handlers):
         # These will hold the values we store from
         # running through the plan steps.
         self._parameters = {}
+        self._current_step_index = -1
+        self._current_step_name = None
+        self._step_names = []
 
     def run(self, plan):
         self._parameters.clear()
-        steps_iter = self._steps(plan)
-        step = next(steps_iter)
+        self._current_step_index = -1
+        self._current_step_name = None
+        self._step_names = list(plan)
+        next_step_name = None
+        step = self._next_step(next_step_name, plan)
         while step is not self._STOP_RUNNING:
             next_step_name = self._run_step(step)
```

Comment:
Would it be better if `_run_step` always returned the next step? Inferring the next step when this value is `None` based off the index doesn't seem great.

Ideally, I think a step should always know what the next step is without the need for this to be tracked via external state. This should be easy to do just by populating the next step field for the implicitly ordered steps when we load the wizard. This removes the need for a `_STOP_RUNNING` sentinel value as `None` would imply there is no next step. This also vastly simplifies the `_next_step` function as there's no longer a difference between an explicit next step and an implicit next step.

Reply:
Capturing offline discussion:

After looking at both versions, I prefer the coroutine version because it keeps all that state local to a single method.

I don't think we should start pre-processing the wizards yet.  The wizard loading happens in a separate object, so we're now creating an implicit contract between what the runner accepts (i.e steps that always have a next step populated) vs what the spec actually supports (implicit next steps).  If anything you could have some sort of pre-processor that's self contained with a specific runner implementation that can fill in missing values (such as next steps) that make the runner logic simpler.  However, the only thing we'd have right now is filling in next steps, and I think it's much simpler to just keep this in a single method instead of introducing some sort of preprocessor object.

If the runner starts getting more complicated then I think we should reevaluate preprocessing.